### PR TITLE
[Bagel]: Support multistage img2img

### PR DESCRIPTION
**Pull request by princepride**
## Purpose

Fully support Bagel multi-stage deploy.

- Patch ModelConfig.is_mm_prefix_lm to include Bagel, enabling bidirectional attention for multimodal prefix positions. Delegates to the original vLLM implementation so upstream list updates are automatically inherited.
- Introduce OmniBagelForConditionalGeneration, wrapping vLLM's Bagel with a built-in VAE encoder for img2img. Supports dual-mode multimodal processing: image modality uses ViT, img2img modality uses VAE. Tracks per-request RoPE and image shape via get_kv_transfer_metadata() for KV transfer to the diffusion stage.
- Add img2img CFG expansion producing 3 companions (user + cfg_text + cfg_img). Update max_batch_size to 3 in bagel.yaml.
- Support custom_metadata in KV cache transfer, allowing model-specific data (RoPE, image shape) to flow from AR to diffusion stage. Propagate received metadata to req.sampling_params.kv_metadata.
- Query model's get_kv_transfer_metadata() to attach custom metadata before KV transfer. Clear model warmup state on first real inference call.

## Test Plan

**img2img(stage0, stage1)**

```
FLASHINFER_DISABLE_VERSION_CHECK=1 python3 examples/offline_inference/bagel/end2end.py   --modality img2img   --image-path women.jpg   --prompts "Let the woman wear a blue dress"
```

bagel_i2i_output

**text2img(stag0, stage1)**

```
FLASHINFER_DISABLE_VERSION_CHECK=1 python3 examples/offline_inference/bagel/end2end.py   --modality text2img   --prompts "A cute cat"
```

image

**text2text(stage0)**

```
FLASHINFER_DISABLE_VERSION_CHECK=1 python3 examples/offline_inference/bagel/end2end.py   --modality text2text   --prompts "Where is the capital of France?"
```

```
'The capital of France is Paris.'
```

**img2text(stage0)**

```
FLASHINFER_DISABLE_VERSION_CHECK=1 python3 examples/offline_inference/bagel/end2end.py   --modality img2text   --image-path text2img.png   --prompts "Please describe this image."
```

```
'This is a digital illustration of a kitten sitting on a wooden floor. The kitten has orange and white fur with darker stripes, large expressive eyes, and pink ears. Its tail is curled around its body, and it appears to be looking directly at the viewer with a curious expression. The background is blurred, suggesting an indoor setting with soft lighting.'
```

### CODE REVIEW

**Comment by princepride** [comment]
@hsliuustc0106 @linyueqian @lishunyang12 PTAL, cc @nussejzz @usberkeley

File: vllm_omni/patch.py [hunk 1]
```diff
@@ -43,6 +45,20 @@ def _patched_glm_image_text_config_init(self, *args, **kwargs):
     # GlmImageTextConfig not available, skip patching
     pass
 
+
+# Patch ModelConfig.is_mm_prefix_lm to include Bagel (bidirectional attention
```
Comment (hsliuustc0106) [inline vllm_omni/patch.py:49]: do we have better ways to avoid such patch?

File: vllm_omni/model_executor/models/bagel/bagel.py [hunk 2]
```diff
@@ -0,0 +1,1009 @@
+from collections import deque
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/bagel/bagel.py:1]: is this file the wapper for the bagel model for AR part?

File: vllm_omni/patch.py (hunk 1 above)
Comment (princepride) [reply vllm_omni/patch.py:49]: We must proceed with this because vllm maintains a list of multi-modal models that require bidirectional attention. We need to integrate Bagel by applying a patch. (Actually, I suspect there might be a bug in my previous Bagel implementation in the vllm; logically, Bagel's mm should also support bidirectional attention.)

File: vllm_omni/model_executor/models/bagel/bagel.py (hunk 2 above)
Comment (princepride) [reply vllm_omni/model_executor/models/bagel/bagel.py:1]: We can't wrapper vLLM's implementation anymore, whether through direct calls or by inheriting modules. First, the vllm's img2text task only requires ViT, whereas our img2img task needs both ViT and VAE. This means we have to modify both the mm_processor and the vision encoder. During implementation, I also discovered another issue: the vision latents for the VAE part actually use DiT model weights rather than AR weights. We must load the full weights into the AR stage.

**Comment by princepride** [comment]
@natureofnature PTAL

**Comment by princepride** [comment]
> do we have any test for this multistage example?

I can add an e2e test for img2img, but it requires an image as input. Previously, in vLLM, Simon helped me upload test images to S3 bucket for CI/CD purposes, but I’m not sure where we are storing our test images in vLLM-Omni now. @congw729

File: vllm_omni/model_executor/models/bagel/bagel.py [hunk 3]
```diff
@@ -0,0 +1,1016 @@
...
+        self.vae2llm = nn.Linear(patch_latent_dim, hidden_size)
+        self.latent_pos_embed = PositionEmbedding(self.max_latent_size, hidden_size)
+        self.time_embedder = TimestepEmbedder(hidden_size)
+
+        self._pending_img2img_info: list[tuple[int, int, int, int]] = []
+        self._ropes_pending: list[dict[str, Any]] = []
+        self._ropes_metadata: dict[str, dict[str, Any]] = {}
+        self._cached_img2img_info: tuple[int, int, int, int] | None = None
+        self._cfg_companions_remaining: int = 0
+
```
Comment (natureofnature) [inline vllm_omni/model_executor/models/bagel/bagel.py:418]: Can it support multiple user request batching? They share the same values and from the yaml file, max batch size is 3, which seems it has a constrain to deal with one img2img request at a time?

Comment (princepride) [reply vllm_omni/model_executor/models/bagel/bagel.py:418]: Can, user can set larger max batch size manually

Comment (natureofnature) [reply vllm_omni/model_executor/models/bagel/bagel.py:418]: > Can, user can set larger max batch size manually

Then they should be changed to dict with request id as keys to avoid data overwritten?

**Comment by congw729** [comment]
> > do we have any test for this multistage example?
> 
> I can add an e2e test for img2img, but it requires an image as input. Previously, in vLLM, Simon helped me upload test images to S3 bucket for CI/CD purposes, but I’m not sure where we are storing our test images in vLLM-Omni now. @congw729

We haven't tried uploading test input images to the S3 bucket. Can you show me the PR/link you just mentioned? I will try to satisfied your need.

**Comment by princepride** [comment]
> > > do we have any test for this multistage example?
> 
> > 
> 
> > I can add an e2e test for img2img, but it requires an image as input. Previously, in vLLM, Simon helped me upload test images to S3 bucket for CI/CD purposes, but I’m not sure where we are storing our test images in vLLM-Omni now. @congw729
> 
> 
> 
> We haven't tried uploading test input images to the S3 bucket. Can you show me the PR/link you just mentioned? I will try to satisfied your need.

https://github.com/vllm-project/vllm/pull/23229#discussion_r2287323064 Last time, Simon Mo helped me upload images to the S3 bucket.

File: vllm_omni/model_executor/models/bagel/bagel.py (hunk 3 above)
Comment (princepride) [reply vllm_omni/model_executor/models/bagel/bagel.py:418]: The suggestion is valid for robustness — if different img2img requests with different resolutions are batched and their CFG companions interleave, the scalar _cached_img2img_info would give wrong dimensions. Converting to a per-request-id dict would fix this edge case. The _ropes_metadata is already correctly keyed by req_id, so it's just _cached_img2img_info and _cfg_companions_remaining that need the same treatment.

File: vllm_omni/patch.py (hunk 1 above)
Comment (princepride) [reply vllm_omni/patch.py:49]: I have raise a pr https://github.com/vllm-project/vllm/pull/36316 let vLLM support bagel vision token support bidirectional attention, in the later version(0.18.0), I will remove this patch.

File: vllm_omni/distributed/omni_connectors/kv_transfer_manager.py [hunk 4]
```diff
@@ -200,8 +200,12 @@ def handle_finished_requests_kv_transfer(
                     logger.warning(f"Request {req_id} has no block IDs, skipping")
                     continue
 
+                custom_metadata = data.get("custom_metadata")
```
Comment (hsliuustc0106) [inline vllm_omni/distributed/omni_connectors/kv_transfer_manager.py:204]: why we need introduce the custom_metadata?  is there any doc to explain this arg? o.w., this many bring poos user/dev experiences.

File: vllm_omni/distributed/omni_connectors/kv_transfer_manager.py [hunk 5]
```diff
@@ -309,14 +316,18 @@ def _normalize_layer_kv(
             Tuple of (key_blocks, value_blocks) if valid, None otherwise
         """
         if isinstance(layer_kv, torch.Tensor):
-            if layer_kv.ndim < 3 or layer_kv.shape[0] != 2:
+            if layer_kv.ndim >= 3 and layer_kv.shape[0] == 2:
```
Comment (hsliuustc0106) [inline vllm_omni/distributed/omni_connectors/kv_transfer_manager.py]: this looks like a kv layer issue? can we make it as a util function?

File: vllm_omni/distributed/omni_connectors/kv_transfer_manager.py (hunk 4 above)
Comment (princepride) [reply vllm_omni/distributed/omni_connectors/kv_transfer_manager.py:204]: Because we need transfer RoPE and image shape from ar stage to dit stage

File: vllm_omni/distributed/omni_connectors/kv_transfer_manager.py (hunk 5 above)
Comment (princepride) [reply vllm_omni/distributed/omni_connectors/kv_transfer_manager.py]: Sure, I will do it. In vLLM, different backend will return different shape of paged attention block, for example in: [`vllm/v1/attention/backends/flash_attn.py`](https://github.com/vllm-project/vllm/blob/1bc9c77f6d324bf7b9253b0c78626fbc50286bfb/vllm/v1/attention/backends/flash_attn.py#L117-L127)
```
@staticmethod
def get_kv_cache_shape(
    num_blocks: int,
    block_size: int,
    num_kv_heads: int,
    head_size: int,
    cache_dtype_str: str = "auto",
) -> tuple[int, ...]:
    if block_size % 16 != 0:
        raise ValueError("Block size must be a multiple of 16.")
    return (2, num_blocks, block_size, num_kv_heads, head_size)
```


However, in: [`vllm/v1/attention/backends/flashinfer.py`](https://github.com/vllm-project/vllm/blob/1bc9c77f6d324bf7b9253b0c78626fbc50286bfb/vllm/v1/attention/backends/flashinfer.py#L318-L326)
```
@staticmethod
def get_kv_cache_shape(
    num_blocks: int,
    block_size: int,
    num_kv_heads: int,
    head_size: int,
    cache_dtype_str: str = "auto",
) -> tuple[int, ...]:
    return (num_blocks, 2, block_size, num_kv_heads, head_size)
```

Comment (princepride) [reply vllm_omni/distributed/omni_connectors/kv_transfer_manager.py]: Done
